### ISSUE TITLE
java.util.ConcurrentModificationException at org.osmdroid.tileprovider.MapTileProviderBase.mapTileRequestFailed(MapTileProviderBase.java:184)

### ISSUE THREAD

**Issue by pasniak** [osmdroid/osmdroid#1446]
## Issue Type

[ ] Question
[x] Bug
## Description and/or steps/code to reproduce the problem

java.util.ConcurrentModificationException
at java.util.LinkedHashMap$LinkedHashIterator.nextEntry(LinkedHashMap.java:346)
at java.util.LinkedHashMap$KeyIterator.next(LinkedHashMap.java:366)
at org.osmdroid.tileprovider.MapTileProviderBase.mapTileRequestFailed(MapTileProviderBase.java:184)
at org.osmdroid.tileprovider.MapTileProviderArray.mapTileRequestFailedExceedsMaxQueueSize(MapTileProviderArray.java:167)
at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$1.removeEldestEntry(MapTileModuleProviderBase.java:124)
at java.util.LinkedHashMap.addNewEntry(LinkedHashMap.java:189)
at java.util.HashMap.put(HashMap.java:404)
at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase.loadMapTileAsync(MapTileModuleProviderBase.java:151)
at org.osmdroid.tileprovider.MapTileProviderArray.runAsyncNextProvider(MapTileProviderArray.java:227)
at org.osmdroid.tileprovider.MapTileProviderArray.mapTileRequestExpiredTile(MapTileProviderArray.java:179)
at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$TileLoader.tileLoadedScaled(MapTileModuleProviderBase.java:299)
at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$TileLoader.run(MapTileModuleProviderBase.java:345)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
at java.lang.Thread.run(Thread.java:818)

My partner was trying my app and the map froze... after a while we got the stack above.  

I am at c245bcd129463a344606f24213acd4471dd5d35c
#906 restoring polyline addPoint function Alex O'Ree 10/26/19 12:27 PM

**Comment by monsieurtanuki** [comment]
@pasniak Is that systematic? I assume that happens at the moment when you add a handler with `MapTileProviderBase.getTileRequestCompleteHandlers().add()`, am I right? Or when you change somehow your activity, like landscape/portrait switch, when you add the handlers in `onCreate`?
I don't think it's linked to #906.

**Comment by pasniak** [comment]
I think this is the first time I see this. We'll check. I do not have a handler added to TileRequestCompleteHandlers (I do  have some commented out code which I totally do not remember why has been added for one layer :-)
``
// tileRequestCompleteHandlers.add(tileRequestCompleteHandler)
``
The rotation is enabled yes but we think not much was happening because the device was on the laying on a desk while the user was touch dragging the map.

**Comment by monsieurtanuki** [comment]
@pasniak Are you sure you don't even use `setTileRequestCompleteHandler`? If not, there's no reason why there should be concurrency.
That being said, something is wrong in the code, as nothing prevents developers to add handlers whenever they want, while the tiles are being downloaded, which can cause concurrency issues.

**Comment by pasniak** [comment]
I only create MapView with context, and I do not break in setTileRequestCompleteHandler. I see SimpleInvalidationHandler attached when mapview is created. It has the MapView ref which it invalidates on Messages with MAPTILE_SUCCESS_ID. It gets attached/detached ok in the lifecycle.

**Comment by spyhunter99** [comment]
@pasniak think you can give this commit a try? i am not too sure how to reproduce it

**Comment by monsieurtanuki** [comment]
Good news: I was able to reproduce the bug, more or less:
````java
new Thread(new Runnable() {
	@Override
	public void run() {
		for (int i = 0 ; i < 10000 ; i ++) {
			mTileRequestCompleteHandlers.add(new TestHandler());
		}
	}
}).start();
new Thread(new Runnable() {
	@Override
	public void run() {
		int messageId = 0;
		for (int i = 0 ; i < 10000 ; i ++) {
			for (final TestHandler handler : mTileRequestCompleteHandlers) {
				if (handler != null) {
					handler.handleMessage(++messageId);
				}
			}
		}
	}
}).start();
````
which produces that kind of log:
````
E/AndroidRuntime: FATAL EXCEPTION: Thread-46181
java.util.ConcurrentModificationException
    at java.util.LinkedHashMap$LinkedHashIterator.nextEntry(LinkedHashMap.java:346)
    at java.util.LinkedHashMap$KeyIterator.next(LinkedHashMap.java:366)
    [in the for-each loop header]
````
Given the context, I believe that this (very rare) issue happens when a new tile is successfully (or not) loaded _at the very same time_ when the `MapView` is created (and the `SimpleInvalidationHandler` added to the handler list).
My bad: I think I introduced this potential (but rare) concurrency issue when I developed #990's `MapSnapshot`s.
I should be able to find a solution this week-end - possibly similar to @spyhunter99's committed suggestion.

**Comment by monsieurtanuki** [comment]
Fixed by #1486.
